server/src/main/java/org/elasticsearch/search/crossproject/CrossProjectIndexResolutionValidator.java

```diff
@@ -143,42 +166,85 @@ public static ElasticsearchException validate(
                 // checking if flat expression matched remotely
                 for (String remoteExpression : remoteExpressions) {
                     String[] splitResource = splitQualifiedResource(remoteExpression);
-                    ElasticsearchException exception = checkSingleRemoteExpression(
+                    var projectAlias = splitResource[0];
+                    var resource = splitResource[1];
+
+                    ElasticsearchException remoteException = checkSingleRemoteExpression(
                         remoteResolvedExpressions,
-                        splitResource[0], // projectAlias
-                        splitResource[1], // resource
+                        projectAlias,
+                        resource,
                         remoteExpression,
                         indicesOptions
                     );
-                    if (exception == null) {
+                    if (remoteException == null) {
                         // found flat expression somewhere
                         foundFlat = true;
+                        if (remoteAuthorizationExceptions != null) {
+                            remoteAuthorizationExceptions.clear();
+                            remoteUnauthorizedIndices.clear();
+                        }
                         break;
                     }
-                    if (unauthorizedException == null && exception instanceof ElasticsearchSecurityException securityException) {
+                    if (unauthorizedException == null && remoteException instanceof ElasticsearchSecurityException securityException) {
                         unauthorizedException = securityException;
+                        if (remoteAuthorizationExceptions == null) {
+                            remoteAuthorizationExceptions = new HashMap<>();
+                            remoteUnauthorizedIndices = new HashMap<>();
+                        }
+                        remoteAuthorizationExceptions.putIfAbsent(projectAlias, securityException);
+                        remoteUnauthorizedIndices.computeIfAbsent(projectAlias, k -> new ArrayList<>()).add(resource);
```

Comment:
Should this be

```suggestion
                        remoteUnauthorizedIndices.computeIfAbsent(projectAlias, k -> new ArrayList<>()).add(remoteExpression);
```
? As discussed over the other PR, the principle is to "use the original expression for local indices and remote expression for remote indices in error reporting"?